File: app/routers/actors.py
"""Actor, version and build-trigger endpoints (Apify /v2/acts + /v2/actors)."""
from __future__ import annotations

from fastapi import APIRouter, Request

from ..auth import resolve_user
from ..responses import data, get_service, not_found, paged_envelope, parse_page, read_json
from ..serializers import actor_dict, build_dict, run_dict, storage_dict, version_dict
from ..service import STORAGE_DS, STORAGE_KV, STORAGE_RQ

# Registered under both /v2/acts and /v2/actors (the CLI uses /v2/actors).
router = APIRouter()

user_router = APIRouter()


async def _my_storages(request: Request, storage_type: str) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    storages = await svc.list_storages_for_user(user, type=storage_type)
    items = [storage_dict(st) for st in storages]
    limit, offset = parse_page(request)
    return data(paged_envelope(items, limit, offset, echo_limit=False, always_total=True))


@user_router.get("/v2/users/me")
async def get_me(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    row = await svc.get_user(user)
    token = row.token if row is not None else None
    return data({"id": user, "username": user, "token": token})


@user_router.get("/v2/users/{user_id_or_username}")
async def get_user_public(user_id_or_username: str, request: Request) -> object:
    """Public profile lookup for ANY user, by id or username.

    Id and username are the same value in this runtime, so one lookup
    serves both. Response is always the public shape (no `token`), even
    for self-lookups. Unknown id/username -> 404 envelope. `resolve_user`
    here is just the bootstrap-or-reject auth guard, not identity resolution.
    """
    svc = get_service(request)
    await resolve_user(request)
    row = await svc.get_user(user_id_or_username)
    if row is None:
        return not_found(f"User '{user_id_or_username}' was not found.")
    return data({"id": row.username, "username": row.username})


@user_router.get("/v2/users/me/actors")
async def my_actors(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    actors = await svc.list_actors(username=user)
    items = [await _actor_payload(svc, a) for a in actors]
    return data({"total": len(items), "count": len(items), "items": items})


@user_router.get("/v2/users/me/builds")
async def my_builds(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    builds = await svc.list_builds_for_user(user)
    items = [build_dict(b) for b in builds]
    return data({"total": len(items), "count": len(items), "items": items})


@user_router.get("/v2/users/me/runs")
async def my_runs(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    runs = await svc.list_runs_for_user(user)
    items = [run_dict(r) for r in runs]
    return data({"total": len(items), "count": len(items), "items": items})


@user_router.get("/v2/users/me/key-value-stores")
async def my_key_value_stores(request: Request) -> object:
    return await _my_storages(request, STORAGE_KV)


@user_router.get("/v2/users/me/datasets")
async def my_datasets(request: Request) -> object:
    return await _my_storages(request, STORAGE_DS)


@user_router.get("/v2/users/me/request-queues")
async def my_request_queues(request: Request) -> object:
    return await _my_storages(request, STORAGE_RQ)


async def _actor_payload(svc, actor) -> dict:
    versions = await svc.list_versions(actor.id)
    tagged = await svc.tagged_builds(actor.id)
    return actor_dict(actor, versions, tagged, svc.settings)


@router.get("")
async def list_actors(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    actors = await svc.list_actors(username=user)
    items = [await _actor_payload(svc, a) for a in actors]
    return data({"total": len(items), "count": len(items), "items": items})


@router.post("")
async def create_actor(request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    body = await read_json(request)
    actor = await svc.create_actor(
        name=body.get("name", "actor"),
        default_run_options=body.get("defaultRunOptions", {}),
        versions=body.get("versions", []),
        username=user,
        actor_standby=body.get("actorStandby"),
    )
    return data(await _actor_payload(svc, actor), status_code=201)


@router.get("/{actor_id}")
async def get_actor(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    actor = await svc.get_actor(actor_id, username=user)
    if actor is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    return data(await _actor_payload(svc, actor))


@router.put("/{actor_id}")
async def update_actor(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    body = await read_json(request)
    actor = await svc.update_actor(actor_id, body if isinstance(body, dict) else {}, username=user)
    if actor is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    return data(await _actor_payload(svc, actor))


@router.get("/{actor_id}/input-schema")
async def get_input_schema(actor_id: str, request: Request) -> object:
    """Resolve the actor's input schema for the console's Input tab.

    Resolved from the SAME version a default (``build=latest``) run would
    actually execute -- the version behind the actor's most recent
    successful build, falling back to its latest-tagged version only when no
    build exists yet (see ``Service.get_input_schema``'s docstring for why).
    Returns ``data(None)`` -- not a 404 -- whenever no schema can be resolved
    (no versions, no manifest/schema file, a TARBALL version, or a malformed
    schema file), matching ``.actor/actor.json``'s own fail-soft inference
    contract. Only an unknown/inaccessible actor id is a 404.
    """
    svc = get_service(request)
    user = await resolve_user(request)
    actor = await svc.get_actor(actor_id, username=user)
    if actor is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    schema = await svc.get_input_schema(actor_id)
    return data(schema)


@router.get("/{actor_id}/versions/{version_number}")
async def get_version(actor_id: str, version_number: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    if await svc.get_actor(actor_id, username=user) is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    version = await svc.get_version(actor_id, version_number)
    if version is None:
        return not_found(f"Version '{version_number}' was not found.")
    return data(version_dict(version))


@router.post("/{actor_id}/versions")
async def create_version(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    if await svc.get_actor(actor_id, username=user) is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    body = await read_json(request)
    version = await svc.upsert_version(actor_id, body)
    return data(version_dict(version), status_code=201)


@router.put("/{actor_id}/versions/{version_number}")
async def update_version(actor_id: str, version_number: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    if await svc.get_actor(actor_id, username=user) is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    body = await read_json(request)
    body.setdefault("versionNumber", version_number)
    version = await svc.upsert_version(actor_id, body)
    return data(version_dict(version))


@router.get("/{actor_id}/builds")
async def list_builds(actor_id: str, request: Request) -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    builds = await svc.list_builds(actor_id, username=user)
    items = [build_dict(b) for b in builds]
    return data({"total": len(items), "count": len(items), "items": items})


@router.post("/{actor_id}/builds")
async def trigger_build(actor_id: str, request: Request, version: str = "0.0", tag: str = "latest") -> object:
    svc = get_service(request)
    user = await resolve_user(request)
    actor = await svc.get_actor(actor_id, username=user)
    if actor is None:
        return not_found(f"Actor '{actor_id}' was not found.")
    ver = await svc.get_version(actor_id, version)
    build_tag = ver.build_tag if ver else tag
    build = await svc.start_build(actor_id, version, build_tag)
    return data(build_dict(build), status_code=201)

File: src/routes/actors.js
/** Actor, version and build-trigger endpoints (Apify /v2/acts + /v2/actors), plus /v2/users/me. */
import { resolveUser } from '../auth.js';
import { STORAGE_DS, STORAGE_KV, STORAGE_RQ } from '../constants.js';
import { data, notFound, readJson } from '../http.js';
import { paginate, parsePage } from '../pagination.js';
import { actorDict, buildDict, runDict, storageDict, versionDict } from '../serializers.js';

function actorPayload(svc, actor) {
    const versions = svc.listVersions(actor.id);
    const tagged = svc.taggedBuilds(actor.id);
    return actorDict(actor, versions, tagged, svc.settings);
}

async function myStorages(ctx, storageType) {
    const svc = ctx.service;
    const user = await resolveUser(ctx);
    const items = svc.listStoragesForUser(user, storageType).map(storageDict);
    const { limit, offset } = parsePage(ctx);
    const page = limit !== null || offset !== null ? paginate(items, limit, offset) : items;
    return data({ total: items.length, count: page.length, items: page });
}

export function registerMeRoutes(router) {
    router.add('GET', '/v2/users/me', async (ctx) => {
        const user = await resolveUser(ctx);
        const row = ctx.service.getUser(user);
        return data({ id: user, username: user, token: row?.token ?? null });
    });

    router.add('GET', '/v2/users/me/actors', async (ctx) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const items = svc.listActors(user).map((a) => actorPayload(svc, a));
        return data({ total: items.length, count: items.length, items });
    });

    router.add('GET', '/v2/users/me/builds', async (ctx) => {
        const user = await resolveUser(ctx);
        const items = ctx.service.listBuildsForUser(user).map(buildDict);
        return data({ total: items.length, count: items.length, items });
    });

    router.add('GET', '/v2/users/me/runs', async (ctx) => {
        const user = await resolveUser(ctx);
        const items = ctx.service.listRunsForUser(user).map(runDict);
        return data({ total: items.length, count: items.length, items });
    });

    router.add('GET', '/v2/users/me/key-value-stores', (ctx) => myStorages(ctx, STORAGE_KV));
    router.add('GET', '/v2/users/me/datasets', (ctx) => myStorages(ctx, STORAGE_DS));
    router.add('GET', '/v2/users/me/request-queues', (ctx) => myStorages(ctx, STORAGE_RQ));

    // Public profile lookup for ANY user, by id or username. Id and username
    // are the same value in this runtime, so one lookup serves both. Response
    // is always the public shape (no `token`), even for self-lookups.
    // `resolveUser` here is just the bootstrap-or-reject auth guard, not
    // identity resolution. Registered AFTER every `/v2/users/me*` route so
    // `me` never resolves as a username.
    router.add('GET', '/v2/users/:userIdOrUsername', async (ctx, { userIdOrUsername }) => {
        await resolveUser(ctx);
        const row = ctx.service.getUser(userIdOrUsername);
        if (!row) return notFound(`User '${userIdOrUsername}' was not found.`);
        return data({ id: row.username, username: row.username });
    });
}

/** Registered under both /v2/acts and /v2/actors (the CLI uses /v2/actors). */
export function registerActorRoutes(router, prefix) {
    router.add('GET', prefix, async (ctx) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const items = svc.listActors(user).map((a) => actorPayload(svc, a));
        return data({ total: items.length, count: items.length, items });
    });

    router.add('POST', prefix, async (ctx) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const body = await readJson(ctx);
        const actor = svc.createActor({
            name: body.name ?? 'actor',
            defaultRunOptions: body.defaultRunOptions ?? {},
            versions: body.versions ?? [],
            username: user,
            actorStandby: body.actorStandby ?? null,
        });
        return data(actorPayload(svc, actor), 201);
    });

    router.add('GET', `${prefix}/:actorId`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const actor = svc.getActor(actorId, user);
        if (!actor) return notFound(`Actor '${actorId}' was not found.`);
        return data(actorPayload(svc, actor));
    });

    router.add('PUT', `${prefix}/:actorId`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const body = await readJson(ctx);
        const payload = body && typeof body === 'object' && !Array.isArray(body) ? body : {};
        const actor = svc.updateActor(actorId, payload, user);
        if (!actor) return notFound(`Actor '${actorId}' was not found.`);
        return data(actorPayload(svc, actor));
    });

    // Resolve the actor's input schema for the console's Input tab. Resolved
    // from the SAME version a default (`build=latest`) run would actually
    // execute -- the version behind the actor's most recent successful
    // build, falling back to its latest-tagged version only when no build
    // exists yet. Returns `data(null)` -- not a 404 -- whenever no schema can
    // be resolved (no versions, no manifest/schema file, a TARBALL version,
    // or a malformed schema file), matching `.actor/actor.json`'s own
    // fail-soft inference contract. Only an unknown/inaccessible actor id is
    // a 404.
    router.add('GET', `${prefix}/:actorId/input-schema`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const actor = svc.getActor(actorId, user);
        if (!actor) return notFound(`Actor '${actorId}' was not found.`);
        return data(svc.getInputSchema(actorId));
    });

    router.add('GET', `${prefix}/:actorId/versions/:versionNumber`, async (ctx, { actorId, versionNumber }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        if (!svc.getActor(actorId, user)) return notFound(`Actor '${actorId}' was not found.`);
        const version = svc.getVersion(actorId, versionNumber);
        if (!version) return notFound(`Version '${versionNumber}' was not found.`);
        return data(versionDict(version));
    });

    router.add('POST', `${prefix}/:actorId/versions`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        if (!svc.getActor(actorId, user)) return notFound(`Actor '${actorId}' was not found.`);
        const body = await readJson(ctx);
        const version = svc.upsertVersion(actorId, body);
        return data(versionDict(version), 201);
    });

    router.add('PUT', `${prefix}/:actorId/versions/:versionNumber`, async (ctx, { actorId, versionNumber }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        if (!svc.getActor(actorId, user)) return notFound(`Actor '${actorId}' was not found.`);
        const body = await readJson(ctx);
        body.versionNumber ??= versionNumber;
        const version = svc.upsertVersion(actorId, body);
        return data(versionDict(version));
    });

    router.add('GET', `${prefix}/:actorId/builds`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const items = svc.listBuilds(actorId, user).map(buildDict);
        return data({ total: items.length, count: items.length, items });
    });

    router.add('POST', `${prefix}/:actorId/builds`, async (ctx, { actorId }) => {
        const svc = ctx.service;
        const user = await resolveUser(ctx);
        const actor = svc.getActor(actorId, user);
        if (!actor) return notFound(`Actor '${actorId}' was not found.`);
        const versionNumber = ctx.query.get('version') || '0.0';
        const tag = ctx.query.get('tag') || 'latest';
        const version = svc.getVersion(actorId, versionNumber);
        const buildTag = version ? version.buildTag : tag;
        const build = svc.startBuild(actorId, versionNumber, buildTag);
        return data(buildDict(build), 201);
    });
}
